# Contents


[1.0. Introduction ](#_Toc86601450)

[2.0 Applications Of AI ](#_Toc86601451)

[3.0. Areas Of AI ](#_Toc86601452)

[4.0. Advantages &amp; Challenges of AI ](#_Toc86601458)

[5.0 Recommender Systems ](#_Toc86601461)

[5.1. What are recommender systems? ](#_Toc86601462)

[5.2. Approaches to recommender systems ](#_Toc86601463)

[5.2.1. Content-Based Approach ](#_Toc86601464)

[5.2.2. Collaborative Approach ](#_Toc86601465)

[6.0. Video Games Recommendation System ](#_Toc86601468)

[8.0. Conclusion ](#_Toc86601471)

[9.0 References ](#_Toc86601472)




# 1.0. Introduction

Artificial intelligence has, in recent years grown to become something that is greatly impacting how society lives day by day. “Artificial Intelligence is probably the most complex and astounding creations of humanity yet.” (Joshi, 2021). From retail, e-commerce, and the media that we consume, artificial intelligence has had an important role to play. Artificial Intelligence is a branch in computer science which has an overall goal of providing systems which are both intelligent and can function independently.  There are many different subdivisions in which Artificial intelligence may lie. Each of which, are applicable to many day-to-day tasks while also being utilised to complete other more comprehensive tasks. 

As part of the continuous assessment for the Artificial Intelligence module, this assignment will consist of a report where research will be conducted with an aim to understand and discuss artificial intelligence and some of its most . With an emphasis on the understanding of ‘Recommender systems’ and how the approach of both collaborative and content-based systems may be applied in such systems. A Recommender system will be created and provided alongside this report using the Python programming language in Jupyter Notebook.


# 2.0 Applications Of AI

There is a wide variety of applications for artificial intelligence around the globe. AI can serve many different purposes and can complete tasks more efficiently with an overall goal to help humans to complete tasks with precision. AI technologies are now being developed to achieve this. Transportation, robotics, and E-Commerce are some of the top sectors utilising AI right now. 

**eCommerce**

eCommerce is the way that merchants interact with the customers that they are dealing with. AI is used in online shopping sites such as amazon to help provide the user with the products they may like based off previous purchasing habits and by  comparing one shoppers’ habit with another’s and providing suggestions from that. 
Amazon has implemented AI into the retrieval of packages by creating parcel retrieval robots named “Ernie” and “Bert”. These move the shelves of items for employees to pick from “so employees don’t have to”. These robots are aware of the whereabouts of each other and can avoid any workplace accidents by always avoiding each other. 

**Media Consumption**

Applications such as Netflix, Spotify and YouTube all used content and collaborative based recommendation systems to help point their end user in the right direction based on their likes, dislikes, or preferences. 

**Transportation**

Transportation is becoming more and more dependent on artificial intelligence as humans find ways to simplify their transport requirements. From self-parking cars to traffic management and monitoring systems which can adapt routes based off current statuses of a route and autonomous driving. One of the leading AI and autopilot developers is Tesla. Tesla work to develop a range of autonomous vehicles and robots with vision to achieve a fully autonomous car. Google and apple are also rumoured to be developing autonomous vehicles with a goal for sale to the public.



# 3.0. Areas Of AI

## 3.1 Machine Learning

Machine learning is a data-based form AI which enables a computer to learn, but when there is enough information provided, the computer can use that additional data to understand the context of that data. Machine learning is a parent topic for the following areas of AI. Machine Learning requires a great amount of data to be able to learn and adapt to suit a particular task. Machine learning has many branching technologies such as:

- Pattern Recognition

- Deep Learning

- Speech Recognition

- NLP

- CNN (Convolutional Neural Networks)

## 3.2. Convolutional neural networks (CNNs)

Convolutional Neural Networks are often found for image recognition and NLP. This uses a grid topology which can recognise and outline objects in an image for the computer to process. This can be applied to object recognition and facial recognition for example. A CNN uses layers which can identify corners and elements of shapes while utilising layers to detect elements of an image. These layers use this information to identify a segment of an image and with more layers, can identify much more of the image.

## 3.3. Natural Language Processing (NLP)

NLP is another branch in AI which is focused on the understanding of text and speech. It uses heavily outlined model-based ruleset to understand the human language. Applications of NLP such as language translation, speech to text and text to speech are heavily used throughout the globe daily.



## 3.4. Computer Vision

Computer vision works in a similar was as regular sight does for humans. The concept of computer vision is to train computers to be able to tell the difference between things like, the distance between objects or whether an object is moving. This is done using cameras, algorithms, and data. A particular industry where this may be applied could be with manufacturing. Computers can be trained to complete a task and become so efficient that its capability to complete that task would surpass that which a human could achieve.

## 3.5. Other forms of AI

Artificial Intelligence may also be considered by the following headings:

**Reactive**

Reactive artificial intelligence has the capability to react to stimulus. It does not have any form of memory. As a result of this, it does not have the ability to learn. It does, however, can respond to combinations of inputs and will not rely on memory driven operations. A good example of an existing artifact of Reactive AI is &#39;Deep Blue&#39; which beat a chess master from learning past chess manoeuvres and skills, which was achieved from building a library of possible outcomes from any given input.

**Limited Memory**

Limited MemoryAI has the capability to learn. It can utilize an allocation of memory to learn and consequentially improve its knowledge. Thereafter it can use its knowledge to respond appropriately.

**Theory Of Mind**

While being able to understand and compute the thoughts and needs of any other existing intelligence, Theory of Mind another form of AI can use great amounts of information of previous events and data to improve upon itself.

**Self-Aware**

An application being self-aware enables the application to have human-like traits and tendencies such as intelligence and self-awareness.

# 4.0. Advantages &amp; Challenges of AI

In terms of strengths and weaknesses of Artificial Intelligence, each can, if not now but in the future outweigh another. There are huge amounts of potential for where AI can take us in the future and how it can help develop the world, we live in. AI will help drive down the time taken to complete a task. All of this is with a positive outlook in the circumstance that AI does not overtake humans and get out of our control. Automation is something that will become a big part of everyday life. Product quality, and a much more efficient use of materials will be something that will be widely used in industries globally.

## Strengths

- Can be applied to many Industries
- Improved social impact at home and in society
- No Human Error
- Cheaper to employ and maintain
- Non-Biased

## Weaknesses

- Outsmarting Humans
- High Cost
- No Imagination &amp; Creativity
- Slow transfer to a new structure

# 5.0 Recommender Systems

# 5.1. What are recommender systems?

Recommender systems are an algorithm which can find and suggest appropriate matches to the end user from a particular dataset provided. This can be either from a content based or collaborative approach. These recommendations can range from social media suggestions, movie, and shopping recommendations. Recommender systems are used widely by big corporations such as Google, Facebook, and Twitter to greater understand their user.

# 5.2. Approaches to recommender systems

There are several different approaches upon creating a recommender system. The two which main examples being Content-Based filtering, and collaborative based filtering.

   ## 5.2.1. Content-Based Approach

This method of recommender system content-based filtering will consider an item&#39;s properties, comparing them to other items properties and proposing appropriate suggestions based on the similarities found between the datasets. This form of recommender system does not store any user data or preferences and can be easily adjusted based on the user&#39;s present preferences. This form of Recommender system requires a great amount of data to be able to provide an accurate output to the user.

   ## 5.2.2. Collaborative Approach

The collaborative approach of AI recommends items based on previous ratings and preferences set from other users in the past which are gathered and presented to the end user. Some common examples of this can be seen with Netflix, Amazon and many other widely used applications.

   ### 5.2.2.1. Memory Based

This form of approach uses past event data from an existing memory often being that of likes, clicks and other user provided data. This establishes correlations between two users as they may like or view similar items. It may also consider other users&#39; data. User based recommender systems are used to recommend an item based on similarities between many users. &quot;Customers that are like you based on X, Y or Z&quot;. This can take age, demographic and similar interests into consideration when suggesting items to the end user. Applications such as Netflix use a previous user interests and ratings to suggest related items to prompt to the end user.

 **Example:** 

Take the following into consideration:

 **Shopper A**  and **Shopper B**  are unrelated. They are both rating products online.

 **Shopper A** has rated  **Product X** 5 stars and  **Product Y** with a 5-star rating.  **Shopper B**  rates  **Product Y**  with a 5-star rating.  **Product X**  would therefore be presented to the  **Shopper B**  based on  **Shopper A&#39;s**  similar interest.

   ### 5.2.2.2. Model Based

A model is a question answering system which would be expected to return a correct answer much of the time. A model is built up on data which is gathered from a dataset. A table can be produced presenting the gathered data from the datasets. The data can then be prepared for use once the data us gathered.

# 6.0. Video Games Recommendation System

## 6.1 Description

This section will run through the steps required to create a functioning recommender system using python which will perform a simple content-based recommendation to the end user. For data, a dataset of video games and their attributes was sourced.  This data is in the form of a ‘.csv’ file sourced online from the Corgis Datasets Project (Corgis, Austin Cory Bart, Dennis Kafura…). The functional goal of this application is to create a recommender system that would suggest appropriate video games to the end user based off the dataset given.  

## 6.2. Implementation

As required for this system to function, a dataset is provided. This dataset contains just over 1200 video game entries and is to be read by the recommender system to provide an appropriate result from the provided recommendation query. It consists of 11 columns each of which providing unique data. This data is represented as [&#39;Title&#39;, &#39;Genre&#39;, &#39;Console&#39; …]

Pandas is a data analysis library that can be added. This is done by importing the library to the current environment. An alias of pandas is created by implicating &#39;as pd&#39; after importing. This enables pandas to be used by typing &#39;pd.function&#39; instead of &#39;pandas.function&#39;.

“Once you have filtered this data, you can analyze it separately and gain insights that might be unique to this group, and inform the predictive modeling steps of the project moving forward” - ("8 Ways To Filter a Pandas DataFrame by a Partial String or Pattern", 2021).

NumPy is another library which stands for Numerical Python, that is added to the application. Like before, NumPy is imported as np instead of NumPy on its own. This library is used to create and analyse data.

In [1]:
import pandas as pd
import numpy as np

In [2]:
games=pd.read_csv(r"video_games.csv")

In [3]:
games

Title  playerNum  Online  \
0             Super Mario 64 DS          1    True   
1        Lumines: Puzzle Fusion          1    True   
2            WarioWare Touched!          2    True   
3      Hot Shots Golf: Open Tee          1    True   
4                  Spider-Man 2          1    True   
..                          ...        ...     ...   
942                     De Blob          4    True   
943  Sonic Riders: Zero Gravity          4    True   
944               Rhythm Heaven          1    True   
945     Tomb Raider: Underworld          1    True   
946  Advance Wars: Days of Ruin          4    True   

                              Genre   Publisher  RS  Sales     UP  \
0                            Action    Nintendo  85   4.69  24.95   
1                          Strategy     Ubisoft  89   0.56  14.95   
2    Action,Racing / Driving,Sports    Nintendo  81   0.54  22.95   
3                            Sports        Sony  81   0.49  12.95   
4                            Action  Activision  61   0.45  14.95   
..                              ...         ...  ..    ...    ...   
942                          Action         THQ  82   0.45  16.95   
943                Racing / Driving        Sega  56   0.45  16.95   
944                          Action    Nintendo  83   0.45  24.95   
945                          Action       Eidos  75   0.44  17.95   
946                        Strategy    Nintendo  86   0.43  17.95   

           Console AgeRating  Year  
0      Nintendo DS         E  2004  
1         Sony PSP         E  2004  
2      Nintendo DS         E  2004  
3         Sony PSP         E  2004  
4      Nintendo DS         E  2004  
..             ...       ...   ...  
942   Nintendo Wii         M  2008  
943   Nintendo Wii         T  2008  
944    Nintendo DS         T  2008  
945  PlayStation 3         T  2008  
946    Nintendo DS         E  2008  

[947 rows x 11 columns]

While dealing with null and duplicate entries, there were a few steps taken to help provide the most efficient game recommendation to the user. &#39;.dropna(inplace=True)&#39; was used to keep any of the game entries with valid entries within the DataFrame, while dropping the invalid entries. &#39;.drop\_duplicates(keep=&quot;first&quot;, inplace=False)&#39; uses the .drop\_duplicates() function to remove any duplicated data objects from the DataFrame. This code was initially excluded from the final code as the dropped rows of data caused complications when requesting a recommendation of a game, as it exceeded the expected bounds of the index of games after the duplicates and null values were removed. However, after some feature testing and trial and error, the issue was resolved by replacing any instances of &#39;_inplace=True_&#39; to &#39;_inplace=False&#39;_. This resolved the issue and games could be suggested even after dropping duplicates and null entities.

&#39;index 946 is out of bounds for axis 0 with size 776&#39;

_fig 1.1 Index Error from asking for a recommendation_

In [4]:
game_sugg = games[['Title', 'Genre', 'Console']]
game_sugg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 947 entries, 0 to 946
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Title    947 non-null    object
 1   Genre    947 non-null    object
 2   Console  947 non-null    object
dtypes: object(3)
memory usage: 22.3+ KB


In [5]:
games.dropna(inplace=False)

Title  playerNum  Online  \
0             Super Mario 64 DS          1    True   
1        Lumines: Puzzle Fusion          1    True   
2            WarioWare Touched!          2    True   
3      Hot Shots Golf: Open Tee          1    True   
4                  Spider-Man 2          1    True   
..                          ...        ...     ...   
942                     De Blob          4    True   
943  Sonic Riders: Zero Gravity          4    True   
944               Rhythm Heaven          1    True   
945     Tomb Raider: Underworld          1    True   
946  Advance Wars: Days of Ruin          4    True   

                              Genre   Publisher  RS  Sales     UP  \
0                            Action    Nintendo  85   4.69  24.95   
1                          Strategy     Ubisoft  89   0.56  14.95   
2    Action,Racing / Driving,Sports    Nintendo  81   0.54  22.95   
3                            Sports        Sony  81   0.49  12.95   
4                            Action  Activision  61   0.45  14.95   
..                              ...         ...  ..    ...    ...   
942                          Action         THQ  82   0.45  16.95   
943                Racing / Driving        Sega  56   0.45  16.95   
944                          Action    Nintendo  83   0.45  24.95   
945                          Action       Eidos  75   0.44  17.95   
946                        Strategy    Nintendo  86   0.43  17.95   

           Console AgeRating  Year  
0      Nintendo DS         E  2004  
1         Sony PSP         E  2004  
2      Nintendo DS         E  2004  
3         Sony PSP         E  2004  
4      Nintendo DS         E  2004  
..             ...       ...   ...  
942   Nintendo Wii         M  2008  
943   Nintendo Wii         T  2008  
944    Nintendo DS         T  2008  
945  PlayStation 3         T  2008  
946    Nintendo DS         E  2008  

[776 rows x 11 columns]

In [6]:
games["Title"]

0               Super Mario 64 DS
1          Lumines: Puzzle Fusion
2              WarioWare Touched!
3        Hot Shots Golf: Open Tee
4                    Spider-Man 2
                  ...            
942                       De Blob
943    Sonic Riders: Zero Gravity
944                 Rhythm Heaven
945       Tomb Raider: Underworld
946    Advance Wars: Days of Ruin
Name: Title, Length: 947, dtype: object

In [7]:
games.isnull().sum()

Title          0
playerNum      0
Online         0
Genre          0
Publisher    171
RS             0
Sales          0
UP             0
Console        0
AgeRating      0
Year           0
dtype: int64

In [8]:
games.dropna(inplace=False)

Title  playerNum  Online  \
0             Super Mario 64 DS          1    True   
1        Lumines: Puzzle Fusion          1    True   
2            WarioWare Touched!          2    True   
3      Hot Shots Golf: Open Tee          1    True   
4                  Spider-Man 2          1    True   
..                          ...        ...     ...   
942                     De Blob          4    True   
943  Sonic Riders: Zero Gravity          4    True   
944               Rhythm Heaven          1    True   
945     Tomb Raider: Underworld          1    True   
946  Advance Wars: Days of Ruin          4    True   

                              Genre   Publisher  RS  Sales     UP  \
0                            Action    Nintendo  85   4.69  24.95   
1                          Strategy     Ubisoft  89   0.56  14.95   
2    Action,Racing / Driving,Sports    Nintendo  81   0.54  22.95   
3                            Sports        Sony  81   0.49  12.95   
4                            Action  Activision  61   0.45  14.95   
..                              ...         ...  ..    ...    ...   
942                          Action         THQ  82   0.45  16.95   
943                Racing / Driving        Sega  56   0.45  16.95   
944                          Action    Nintendo  83   0.45  24.95   
945                          Action       Eidos  75   0.44  17.95   
946                        Strategy    Nintendo  86   0.43  17.95   

           Console AgeRating  Year  
0      Nintendo DS         E  2004  
1         Sony PSP         E  2004  
2      Nintendo DS         E  2004  
3         Sony PSP         E  2004  
4      Nintendo DS         E  2004  
..             ...       ...   ...  
942   Nintendo Wii         M  2008  
943   Nintendo Wii         T  2008  
944    Nintendo DS         T  2008  
945  PlayStation 3         T  2008  
946    Nintendo DS         E  2008  

[776 rows x 11 columns]

In [9]:
games.duplicated().value_counts()

False    947
dtype: int64

In [10]:
games=games.drop_duplicates(keep="first", inplace=False)

The sklearn.feature library is used to import the &#39;TfidfVectorizer&#39;. This is the beginning of the algorithm that recommends to the user. The &#39;tfv\_matrix&#39; is also included. This evaluates the relevance of the words in the dataset and determines their importance which is used in NLP. This matrix enables the data to be filtered by title.



In [11]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [12]:
vector_game = TfidfVectorizer(min_df = 3,
                         max_features = None,
                         strip_accents = 'unicode',
                         analyzer = 'word',
                         token_pattern = '\w{2,}',
                         ngram_range = (1,3),
                         stop_words = 'english')

In [13]:
tfv_matrix=vector_game.fit_transform(games["Title"])
#after recording the video i remembered this applies the parameters to the game inputted
tfv_matrix.shape

(947, 518)

In [14]:
from sklearn.metrics.pairwise import sigmoid_kernel

&#39;sigmoid\_kernal&#39; is originated from Neural Networks. It is an algorithm used for the analysis of the data provided. A sigmoid is known as a hyperbolic tangent which determines an output for each object of the dataset.

![](RackMultipart20211031-4-65sfw1_html_9438f49dcdcdc540.gif)

 Fig 1.1 “hyperbolic tangent function for real values of its argument” ("Hyperbolic tangent: Introduction to the Hyperbolic Tangent Function", 2021)


In [15]:
sig_game=sigmoid_kernel(tfv_matrix, tfv_matrix)

In [16]:
sig_game

array([[0.76240373, 0.76159416, 0.76159416, ..., 0.76159416, 0.76159416,
        0.76159416],
       [0.76159416, 0.76240373, 0.76159416, ..., 0.76159416, 0.76159416,
        0.76159416],
       [0.76159416, 0.76159416, 0.76159416, ..., 0.76159416, 0.76159416,
        0.76159416],
       ...,
       [0.76159416, 0.76159416, 0.76159416, ..., 0.76159416, 0.76159416,
        0.76159416],
       [0.76159416, 0.76159416, 0.76159416, ..., 0.76159416, 0.76240373,
        0.76159416],
       [0.76159416, 0.76159416, 0.76159416, ..., 0.76159416, 0.76159416,
        0.76240373]])

In [17]:
sig_game.shape

(947, 947)

In [18]:
# index = games.reset_index()

In [23]:
index = pd.Series(games.index, index=games['Title']).drop_duplicates()
# index = pd.Series(games.index, index=games['Title'])

In [20]:
# index.reset_index()

The final stage of the setup of the recommendation system is to create the similarity score function. A new function, &#39;recommend\_game&#39; is created which will be used to run the recommendation. Two variables are set within this function – title and sig\_kern (which is redefined from sig\_game from the previous chunk of code.) &#39;indx&#39; is defined as the index of the game from the dataset (e.g., &quot;Madden NFL 2005&quot; has the index of &#39;8&#39;).

The line &#39;sigmoid\_score = sorted(sigmoid\_score, key = lambda x:x[1], reverse = True)&#39; simply sorts the data in the CSV file.

&#39;sigmoid\_score[1:4]&#39; determines the number of suggested video games presented to the user.

In [24]:
def recommend_game(title, sig_kern=sig_game):
    indx = index[title]

    sigmoid_score = list(enumerate(sig_kern[indx]))
    sigmoid_score = sorted(sigmoid_score, key = lambda x:x[1], reverse = True)
    sigmoid_score = sigmoid_score[1:4]
    game_found = [i[0] for i in sigmoid_score]
    return games.iloc[game_found]

A Final Linne of code may be run to input a name of a video game to the system with the expected output of three suggested games based off the algorithm detailed above. This code is represented as "recommend_game('game_title_here')"

In [25]:
recommend_game('Rhythm Heaven')

Title  playerNum  Online  \
1    Lumines: Puzzle Fusion          1    True   
2        WarioWare Touched!          2    True   
3  Hot Shots Golf: Open Tee          1    True   

                            Genre Publisher  RS  Sales     UP      Console  \
1                        Strategy   Ubisoft  89   0.56  14.95     Sony PSP   
2  Action,Racing / Driving,Sports  Nintendo  81   0.54  22.95  Nintendo DS   
3                          Sports      Sony  81   0.49  12.95     Sony PSP   

  AgeRating  Year  
1         E  2004  
2         E  2004  
3         E  2004

In [26]:
recommend_game('Advance Wars: Days of Ruin')

Title  playerNum  Online                    Genre  \
699  Geometry Wars: Galaxies          1    True                   Action   
709  Geometry Wars: Galaxies          2    True                   Action   
727    Crazy Taxi: Fare Wars          1    True  Action,Racing / Driving   

    Publisher  RS  Sales     UP       Console AgeRating  Year  
699       NaN  79   0.12  14.95   Nintendo DS         T  2007  
709       NaN  80   0.11  16.95  Nintendo Wii         T  2007  
727      Sega  64   0.10  17.95      Sony PSP         E  2007


An attempt to solve an error of the following output was conducted: &quot;The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()&quot; which is caused by duplicate entries of video games in the database being present. The method of rectification of this error was to run an if statement which followed the following pseudo code:

 **If** there no duplicates **then** &#39;sigmoid\_score = list(enumerate(sig\_kern[indx]))&#39;,

 **Else** , &#39;sigmoid\_score = list(enumerate(sig\_kern[indx[0]]))&#39;.

This code would either expect a single response within the index or, it would expect an array, being the duplicates in the else statement. However, this functionality was not functional upon achieving the assessment deadline.

# 8.0. Conclusion

There is so much more to come from Artificial Intelligence in the future. With our daily interactions with AI rising day to day, its only a matter of time when it is applicable to every element of societies development. E-commerce will see a huge incline in dependencies with AI and many governments will begin to adopt this as a new way to further their systems. From the time spent on the creation and research of recommender systems, it can be determined that this is a great entry point into the advancement and interpretation of Artificial Intelligence.

# References

(Anon, 2021). Retrieved 30 October 2021, from https://www.cnbc.com/2021/06/13/amazon-details-new-warehouse-robots-ernie-and-bert.html

Artificial Intelligence & Autopilot. (2021). Retrieved 30 October 2021, from https://www.tesla.com/AI

Artificial Intelligence - Research Areas. (2021). Retrieved 25 October 2021, from https://www.tutorialspoint.com/artificial_intelligence/artificial_intelligence_research_areas.htm

CORGIS Datasets Project. (2021). Retrieved 14 October 2021, from https://corgis-edu.github.io/corgis/csv/video_games/

Hyperbolic tangent: Introduction to the Hyperbolic Tangent Function. (2021). Retrieved 20 October 2021, from https://functions.wolfram.com/ElementaryFunctions/Tanh/introductions/Tanh/ShowAll.html#:~:text=The%20hyperbolic%20tangent%20function%20is%20an%20old%20mathematical,two%20exponential%20functions%20in%20the%20points%20and%20%29%3A

Joshi, N. (2021). 7 Types Of Artificial Intelligence. Retrieved 20 October 2021, from https://www.forbes.com/sites/cognitiveworld/2019/06/19/7-types-of-artificial-intelligence/?sh=a37a559233ee

Recommendation system in Python. (2021). Retrieved 26 October 2021, from https://medium.com/dataseries/recommendation-system-in-python-4d4cf6d33166

8 Ways To Filter a Pandas DataFrame by a Partial String or Pattern. (2021). Retrieved 14 October 2021, from https://towardsdatascience.com/8-ways-to-filter-a-pandas-dataframe-by-a-partial-string-or-pattern-49f43279c50f


Simple Movie Recommendation System | Python | TF-IDF Vectorizer | NLP. Retrieved 14 October 2021, from https://www.youtube.com/watch?v=rzxket9UtL4&t=811s&ab_channel=CodeArt
